### PR TITLE
NAS-111838 / 21.10 / Move LDAP client code into explicit service plugin

### PR DESCRIPTION
**Pull request by anodos325**
This helps to standardize LDAP access and provides a somewhat
more abstract mechanism for interacting with remote LDAP servers
(if needed for AD plugin or other parts of middleware).

### CODE REVIEW

**Review by yocalebo — CHANGES_REQUESTED** [review]
`_search` doesn't have `timeout` set which means this can block forever if the called method never returns.

**Review by yocalebo — APPROVED** [review]
My only concern is the main interpreter hanging during restart/shutdown if this thread is created and running before the timeout has elapsed.

Granted, it's probably not appropriate for you to solve that problem here since it requires adding some type of mechanism (threading event) that's shared between this ldap thread and parent thread. Anyways, other than that looks fine to me.

File: src/middlewared/middlewared/plugins/ldap.py [hunk 1]
```diff
@@ -348,7 +348,7 @@ def _search(self, ldap_config, basedn='', scope=pyldap.SCOPE_SUBTREE, filter='',
                     attrsonly=0,
                     serverctrls=serverctrls,
                     clientctrls=clientctrls,
-                    timeout=timeout,
+                    timeout=ldap_config['options']['timeout'],
```
Comment (yocalebo) [inline src/middlewared/middlewared/plugins/ldap.py:351]: Actually, wondering if we should make sure this is some sane value (instead of like 9999999).

**Review by yocalebo — CHANGES_REQUESTED** [review]

**Comment by rick-mesta** [comment]
@anodos325: ping -- is there any of @yocalebo 's comments that may require additional changes ?

**Review by yocalebo — APPROVED** [review]

**Pull request merged**